### PR TITLE
Try to set up CASE session over all discovered ips instead of only th…

### PR DESCRIPTION
**Pull request by vivien-apple**
…e 'best' one that has been discovered

#### Problem

This is the second part of #22789 that was addressing `PASE` session establishment. The same issue happens for CASE where  that some of the discovered ips from mdns are staled cache entry.

### CODE REVIEW

**Comment by vivien-apple** [comment]
I guess I need to fix the BSS size increase :(

File: src/app/OperationalSessionSetup.cpp [hunk 1]
```diff
@@ -420,6 +435,28 @@ void OperationalSessionSetup::PerformAddressUpdate()
...
+        if (currentAddress == result.address.GetIPAddress())
+        {
+            continue;
+        }
+
+        ResolveResult * resolveResult = mDiscoveredResults.CreateObject();
+        if (resolveResult == nullptr)
+        {
+            // It may happens if there are more results than the available number of slots in the pool. This
+            // is not fatal but it may indicate that the pool size must be increase.
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:452]:
```suggestion
            // It may happen if there are more results than the available number of slots in the pool. This
            // is not fatal but it may indicate that the pool size must be increased.
```

File: src/app/OperationalSessionSetup.cpp [hunk 2]
```diff
@@ -420,6 +435,28 @@ void OperationalSessionSetup::PerformAddressUpdate()
...
+            ChipLogError(Discovery, "One discovered IP will not be tried during Operational Session Setup");
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:453]:
```suggestion
            ChipLogError(Discovery, "Some discovered IPs will not be tried during Operational Session Setup");
```

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp [hunk 3]
```diff
@@ -180,6 +181,26 @@ void Resolver::Shutdown()
...
 {
+    ResolveResult result;
+
+    auto resolutionData = nodeData.resolutionData;
+    result.address.SetPort(resolutionData.port);
+    result.address.SetInterface(resolutionData.interfaceId);
+    result.mrpRemoteConfig = resolutionData.GetRemoteMRPConfig();
+    result.supportsTcp     = resolutionData.supportsTcp;
+
+    // Copy all the IPs that has been found and sort them by score.
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:192]:
```suggestion
    // Copy all the IPs that have been found and sort them by score.
```

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp [hunk 4]
```diff
@@ -180,6 +181,26 @@ void Resolver::Shutdown()
 
 void Resolver::OnOperationalNodeResolved(const Dnssd::ResolvedNodeData & nodeData)
 {
+    ResolveResult result;
+
+    auto resolutionData = nodeData.resolutionData;
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:186]: Does this have to copy, or could it be `auto &`?

File: src/app/OperationalSessionSetup.h [hunk 5]
```diff
@@ -326,6 +327,8 @@ class DLL_EXPORT OperationalSessionSetup : public SessionDelegate,
      * This function will set new IP address, port and MRP retransmission intervals of the device.
      */
     void UpdateDeviceData(const Transport::PeerAddress & addr, const ReliableMessageProtocolConfig & config);
+
+    ObjectPool<AddressResolve::ResolveResult, Dnssd::CommonResolutionData::kMaxIPAddresses> mDiscoveredResults;
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.h:331]: I suspect the BSS problem is that ResolveResult is not that small, we have a pool of them here, and we have a pool of OperationalSessionSetup objects, so we end up with lots of these ResolveResults allocated....

How much RAM are we willing to spend on this?  Unlike commissionable discovery, operational discovery has to be done by constrained devices...

Comment (vivien-apple) [reply src/app/OperationalSessionSetup.h]: I don't need to hold "ResolveResult". I have changed the code with a struct that is smaller:
```
struct DeviceData
{
    Transport::PeerAddress address;
    ReliableMessageProtocolConfig mrpRemoteConfig;

    DeviceData() : address(Transport::Type::kUdp), mrpRemoteConfig(GetDefaultMRPConfig()) {}
};

ObjectPool<DeviceData, Dnssd::CommonResolutionData::kMaxIPAddresses> mDiscoveredDeviceData;
```

That said, I still suspect the BSS increase to be fairly large. We won't hold a pool of struct that contains a pool, but the first pool is size `64` by default https://github.com/project-chip/connectedhomeip/blob/f54f0f6a46f849fecafc26bb44652a5c4f0f0fec/src/lib/core/CHIPConfig.h#L946 which is already a lot for constrained device.

Comment (vivien-apple) [reply src/app/OperationalSessionSetup.h]: Ah I have read the wrong macro. It should be `4` by default for those device, which is much more reasonable.
```
https://github.com/project-chip/connectedhomeip/blob/f54f0f6a46f849fecafc26bb44652a5c4f0f0fec/src/lib/core/CHIPConfig.h#L983
```

Comment (vivien-apple) [reply src/app/OperationalSessionSetup.h]: I have updated the code such that it works a bit differently. The reason behind this change is because I found that a `ResolveResult` struct was already defined in `mAddressLookupHandle` which was resulting into doubling the additional memory.
So, instead of making a local copy on `src/app/OperationalSessionSetup.*`, the address lookup handle keeps a list of results (configurable by a flag) and returns them as needed.

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp [hunk 6]
```diff
@@ -30,41 +30,28 @@ void NodeLookupHandle::ResetForLookup(System::Clock::Timestamp now, const NodeLo
...
 
-        if (!mBestResult.address.GetIPAddress().IsIPv6LinkLocal())
-        {
-            // Only use the DNS-SD resolution's InterfaceID for addresses that are IPv6 LLA.
-            // For all other addresses, we should rely on the device's routing table to route messages sent.
-            // Forcing messages down an InterfaceId might fail. For example, in bridged networks like Thread,
-            // mDNS advertisements are not usually received on the same interface the peer is reachable on.
-            mBestResult.address.SetInterface(Inet::InterfaceId::Null());
-            ChipLogDetail(Discovery, "Lookup clearing interface for non LL address");
-        }
```
Comment (tcarmelveilleux) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:58]: Why are you losing this? This is very problematic.

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp [hunk 7]
```diff
@@ -30,41 +30,28 @@ void NodeLookupHandle::ResetForLookup(System::Clock::Timestamp now, const NodeLo
...
+    auto score = Dnssd::IPAddressSorter::ScoreIpAddress(result.address.GetIPAddress(), result.address.GetInterface());
```
Comment (tcarmelveilleux) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:42]: UpdateResult below immediately calls `ScoreIpAddress`. Suggest a way that doesn't force the score to be computed twice

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp [hunk 8]
```diff
@@ -112,6 +100,65 @@ NodeLookupAction NodeLookupHandle::NextAction(System::Clock::Timestamp now)
...
+    // Move the following valid entries one level down.
+    for (auto i = count; i > insertAtIndex; i--)
+    {
+        if (i >= kNodeLookupResultsLen)
+        {
+            continue;
+        }
+
+        results[i] = results[i - 1];
+    }
```
Comment (tcarmelveilleux) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:139]: Please add unit tests to guarantee this code works. Whenever you have code that moves entries in place, there is potential for indexing mistakes. I looked hard at the indexing, and I can't convince myself there isn't a corner case or off-by-one in the method

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp [hunk 9]
```diff
@@ -112,6 +100,65 @@ NodeLookupAction NodeLookupHandle::NextAction(System::Clock::Timestamp now)
...
+    if (count < kNodeLookupResultsLen)
+    {
+        count++;
+    }
+
+    auto & updatedResult = results[insertAtIndex];
+    updatedResult        = result;
+    if (!updatedResult.address.GetIPAddress().IsIPv6LinkLocal())
+    {
+        // Only use the DNS-SD resolution's InterfaceID for addresses that are IPv6 LLA.
```
Comment (ksperling-apple) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:151]: This seems like it maybe should be common logic that lives in the resolver itself?

Comment (vivien-apple) [reply src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:143]: I am not disagreeing but I'm not sure if some implementation needs to not filter that and that is why it lives in the "DefaultImpl" that can be overridden.

File: src/lib/address_resolve/tests/TestAddressResolve_DefaultImpl.cpp [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    Inet::IPAddress::FromString("::/80", ipAddress);
+
+    return Transport::PeerAddress::UDP(ipAddress, port, interfaceId);
+}
+
+const Transport::PeerAddress GetAddressWithMediumScore(uint16_t port                 = CHIP_PORT,
+                                                       Inet::InterfaceId interfaceId = Inet::InterfaceId::Null())
+{
+    Inet::IPAddress ipAddress;
+    Inet::IPAddress::FromString("192.168.0.1", ipAddress);
```
Comment (tcarmelveilleux) [inline src/lib/address_resolve/tests/TestAddressResolve_DefaultImpl.cpp:43]: IPv4 addresses should not be considered in this test unless IPv4 support is enabled. Please remove IPv4 example.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp (hunk 9 above)
Comment (bzbarsky-apple) [reply src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:143]: "the resolver" is actually multiple implementations, so in some ways this is the right place to have the logic if it's going to only live in one place....  Note also that this is pre-existing.

File: src/lib/address_resolve/AddressResolve.h [hunk 11]
```diff
@@ -204,6 +204,21 @@ class Resolver
...
+    /// If this returns CHIP_NO_ERROR, the following is expected:
+    ///   - The listener OnNodeAddressResolved will be called with the additional data.
+    ///   - handle must NOT be destroyed while the lookup is in progress (it
+    ///     is part of an internal 'lookup list')
+    ///   - handle must NOT be reused (the lookup is done on a per-node basis
+    ///     and maintains lookup data internally while the operation is still
+    ///     in progress)
+    ///
+    /// If no additional data is available at the time of the request, it returns HIP_ERROR_WELL_EMPTY.
+    virtual CHIP_ERROR LookupNode(Impl::NodeLookupHandle & handle) = 0;
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve.h:220]: This feels like it should really be called `TryNextResult` or something like that.

File: src/lib/address_resolve/AddressResolve_DefaultImpl.h [hunk 12]
```diff
@@ -26,13 +26,44 @@ namespace chip {
...
+    uint8_t consumed                             = 0; // number of already read ResolveResult
+
+    bool UpdateResults(const ResolveResult & result, Dnssd::IPAddressSorter::IpScore score);
+
+    bool HasValidResult() const { return count > consumed; }
+
+    ResolveResult ConsumeResult()
+    {
+        VerifyOrDie(HasValidResult());
+        return results[consumed++];
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.h:51]: This part is not making sense to me, unless I am missing something.

Let's say we get two results, A and B.  Array is [A, B].  We consume A then a new result C comes in with a better score.  The array ends up as [C, A, B], `consumed` is set to 1, so the next time we will try A again, then try B, and never try C.

Fundamentally, since new results can get inserted in between existing possibly-consumed ones with our scoring setup a single "consumed" counter doesn't seem to be able to represent the state of things.

I see two possible solutions offhand:
1. ConsumeResult in fact moves things over by one and reduces `count`.
2. We change the ordering of things in the array so smaller scores come first.  That makes it cheap to add a new thing with a new high score (no shifting needed), and cheap to pop off the thing with the best score (just decrement count), but makes adding a new low-scoring thing more expensive (need to shift, and need to be careful to not shift out high-scoring things).

Either way, `consumed` would go away.

Looking at the tests, they don't seem to test these "interleave consumption with addition" scenarios, and probably should.

File: src/lib/address_resolve/AddressResolve_DefaultImpl.h [hunk 13]
```diff
@@ -116,15 +147,20 @@ class NodeLookupHandle : public NodeLookupHandleBase
     /// any active searches)
     NodeLookupAction NextAction(System::Clock::Timestamp now);
 
+    /// Does the node has any valid lookup result
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.h:150]:
```suggestion
    /// Does the node have any valid lookup result?
```

File: src/lib/address_resolve/AddressResolve_DefaultImpl.h [hunk 14]
```diff
@@ -116,15 +147,20 @@ class NodeLookupHandle : public NodeLookupHandleBase
...
+    bool HasLookupResult() const { return mResults.HasValidResult(); }
+
+    /// Return the next valid lookup result.
+    ResolveResult GetLookupResult() { return mResults.ConsumeResult(); }
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.h:154]: Feels like this should be called `TakeLookupResult` to make it clear that it advances the state.

File: src/lib/core/CHIPConfig.h [hunk 15]
```diff
@@ -1270,6 +1270,17 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
 #define CHIP_CONFIG_MINMDNS_MAX_PARALLEL_RESOLVES 2
 #endif // CHIP_CONFIG_MINMDNS_MAX_PARALLEL_RESOLVES
 
+/**
+ * def CHIP_CONFIG_MDNS_RESOLVE_LOOKUP_RESULTS
+ *
+ * @brief Determines the maximum number of node resolve informations to keep
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPConfig.h:1276]:
```suggestion
 * @brief Determines the maximum number of node resolve results (PeerAddresses) to keep
```

File: src/lib/address_resolve/AddressResolve_DefaultImpl.h (hunk 12 above)
Comment (vivien-apple) [reply src/lib/address_resolve/AddressResolve_DefaultImpl.h:51]: As discussed on slack I have added a check in `TryNextResult` to make it clearer that the lookup handle is not expected to be active when it is called. As you said that would result into all sort of weird results.

**Comment by vivien-apple** [comment]
It seems like the EFR32 build does not like one of the way I have formatted a `ChipLogProgress`, it fails with: `FLASH memory overlapped with NVM section`
I will revert this change.

**Review by woody-apple — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]
